File: android-native/app/src/main/java/flexr/social/app/data/repository/Mappers.kt
package flexr.social.app.data.repository

import flexr.social.app.core.common.ServerTime
import flexr.social.app.data.remote.dto.AgeCheckResponseDto
import flexr.social.app.data.remote.dto.GymDto
import flexr.social.app.data.remote.dto.MatchDto
import flexr.social.app.data.remote.dto.MembershipStatusDto
import flexr.social.app.data.remote.dto.MessageDto
import flexr.social.app.data.remote.dto.MyProfileDto
import flexr.social.app.data.remote.dto.PhotoDto
import flexr.social.app.data.remote.dto.ProfileDto
import flexr.social.app.data.remote.dto.PushNotificationDto
import flexr.social.app.data.remote.dto.VerificationStatusDto
import flexr.social.app.domain.model.AgeCheck
import flexr.social.app.domain.model.Gender
import flexr.social.app.domain.model.Gym
import flexr.social.app.domain.model.MatchSummary
import flexr.social.app.domain.model.Membership
import flexr.social.app.domain.model.Message
import flexr.social.app.domain.model.MyProfile
import flexr.social.app.domain.model.Photo
import flexr.social.app.domain.model.PhotoStatus
import flexr.social.app.domain.model.NotificationSettings
import flexr.social.app.domain.model.Profile
import flexr.social.app.domain.model.PushNotification
import flexr.social.app.domain.model.VerificationDocumentType
import flexr.social.app.domain.model.VerificationState
import flexr.social.app.domain.model.VerificationStatus
import flexr.social.app.domain.model.VerificationStep
import java.time.Instant

/** DTO → Domain. Bewusst an einer Stelle gebündelt, damit die Übersetzung des
 *  Backend-Vertrags nachvollziehbar bleibt. */

fun PhotoDto.toDomain() = Photo(
    id = id,
    url = url,
    thumbUrl = thumbUrl,
    position = position,
    status = PhotoStatus.from(status),
)

fun ProfileDto.toDomain() = Profile(
    id = id,
    name = name,
    age = age,
    city = city,
    gender = Gender.from(gender),
    gym = gym,
    bio = bio,
    isOnline = isOnline,
    isVerified = isVerified,
    distanceKm = distanceKm,
    photos = photos.sortedBy { it.position }.map { it.toDomain() },
)

fun MyProfileDto.toDomain() = MyProfile(
    profile = Profile(
        id = id,
        name = name,
        age = age,
        city = city,
        gender = Gender.from(gender),
        gym = gym,
        bio = bio,
        isOnline = isOnline,
        isVerified = isVerified,
        distanceKm = distanceKm,
        // Die eigene Ansicht zeigt bewusst ALLE Fotos inklusive Moderationsstatus.
        photos = photos.sortedBy { it.position }.map { it.toDomain() },
    ),
    plz = plz,
    birthdate = ServerTime.parseDate(birthdate),
    searchRadiusKm = searchRadiusKm,
    messagingMutedUntil = ServerTime.parse(messagingMutedUntil),
    verificationRequired = verificationRequired,
    isAccountActivated = isAccountActivated,
    email = email,
    emailVerified = emailVerified,
    ageVerified = ageVerified,
    notifications = NotificationSettings(
        matchEmail = notifyMatchEmail,
        matchPush = notifyMatchPush,
        queueEmail = notifyQueueEmail,
        queuePush = notifyQueuePush,
        inactiveEmail = notifyInactiveEmail,
        inactivePush = notifyInactivePush,
    ),
)

fun PushNotificationDto.toDomain() = PushNotification(
    id = id,
    topic = topic,
    title = title,
    body = body,
    target = target,
)

fun MembershipStatusDto.toDomain() = Membership(
    isSubscribed = isSubscribed,
    trialEndsAt = ServerTime.parse(trialEndsAt) ?: Instant.EPOCH,
    isActive = isActive,
)

fun MessageDto.toDomain() = Message(
    id = id,
    matchId = matchId,
    senderId = senderId,
    content = content,
    createdAt = ServerTime.parse(createdAt) ?: Instant.now(),
    readAt = ServerTime.parse(readAt),
    wasCensored = wasCensored,
)

fun MatchDto.toDomain() = MatchSummary(
    matchId = matchId,
    profile = profile.toDomain(),
    lastMessage = lastMessage?.toDomain(),
    unreadCount = unreadCount,
    isOnline = isOnline,
    inChats = inChats,
)

fun GymDto.toDomain() = Gym(
    id = id,
    name = name,
    street = street,
    houseNumber = houseNumber,
    plz = plz,
    city = city,
    label = label,
)

fun VerificationStatusDto.toDomain() = VerificationState(
    status = VerificationStatus.from(status),
    prompts = prompts.orEmpty(),
    nextStep = VerificationStep.from(nextStep),
    reason = reason,
    verificationRequired = verificationRequired,
    accountActivated = accountActivated,
    emailVerified = emailVerified,
    documentTypes = documentTypes.orEmpty().map {
        VerificationDocumentType(value = it.value, label = it.label, needsBack = it.needsBack)
    },
)

fun AgeCheckResponseDto.toDomain() = AgeCheck(
    eligible = eligible,
    age = age,
    message = message,
)

File: ios/FLEXR/Data/Repository/Mappers.swift
import Foundation

/// DTO → Domäne. Bewusst an einer Stelle gebündelt, damit die Übersetzung des
/// Backend-Vertrags nachvollziehbar bleibt.

extension PhotoDTO {
    func toDomain() -> Photo {
        Photo(
            id: id,
            url: url,
            thumbURL: thumbUrl,
            position: position ?? 0,
            status: PhotoStatus(raw: status)
        )
    }
}

extension ProfileDTO {
    func toDomain() -> Profile {
        Profile(
            id: id,
            name: name,
            age: age,
            city: city,
            gender: Gender(raw: gender),
            gym: gym,
            bio: bio,
            isOnline: isOnline ?? false,
            isVerified: isVerified ?? false,
            isPremium: isPremium ?? false,
            distanceKm: distanceKm,
            photos: (photos ?? []).sorted { $0.position ?? 0 < $1.position ?? 0 }.map { $0.toDomain() }
        )
    }
}

extension IncomingLikesDTO {
    func toDomain() -> IncomingLikes {
        IncomingLikes(
            count: count ?? 0,
            profiles: (profiles ?? []).map { $0.toDomain() },
            premiumRequired: premiumRequired ?? false
        )
    }
}

extension MyProfileDTO {
    func toDomain() -> MyProfile {
        MyProfile(
            profile: Profile(
                id: id,
                name: name,
                age: age,
                city: city,
                gender: Gender(raw: gender),
                gym: gym,
                bio: bio,
                isOnline: isOnline ?? false,
                isVerified: isVerified ?? false,
                isPremium: isPremium ?? false,
                distanceKm: distanceKm,
                // Die eigene Ansicht zeigt bewusst ALLE Fotos inklusive Moderationsstatus.
                photos: (photos ?? [])
                    .sorted { $0.position ?? 0 < $1.position ?? 0 }
                    .map { $0.toDomain() }
            ),
            plz: plz,
            birthdate: ServerTime.parseDate(birthdate),
            searchRadiusKm: searchRadiusKm ?? 20,
            messagingMutedUntil: ServerTime.parse(messagingMutedUntil),
            email: email ?? "",
            emailVerified: emailVerified ?? true,
            verificationRequired: verificationRequired ?? false,
            isAccountActivated: isAccountActivated ?? true,
            ageVerified: ageVerified ?? false,
            notifications: NotificationSettings(
                matchEmail: notifyMatchEmail ?? true,
                matchPush: notifyMatchPush ?? true,
                queueEmail: notifyQueueEmail ?? true,
                queuePush: notifyQueuePush ?? true,
                inactiveEmail: notifyInactiveEmail ?? true,
                inactivePush: notifyInactivePush ?? true,
                pendingLikesEmail: notifyPendingLikesEmail ?? true,
                pendingLikesPush: notifyPendingLikesPush ?? true
            ),
            language: language ?? "de"
        )
    }
}

extension MembershipStatusDTO {
    func toDomain() -> Membership {
        // Die Standardwerte gelten nur, wenn der Server ein Feld gar nicht
        // liefert. Sie beschreiben bewusst den zurückhaltendsten Fall: Premium
        // nicht kaufbar, keine Grenzen aktiv, Beta-Abzeichen an.
        //
        // `checkoutAvailable` hat Vorrang vor `premiumEnabled`; ein Server, der
        // es noch nicht kennt, fällt auf `premiumEnabled` zurück. Beide sind
        // für diese App ohnehin falsch, solange der Verkauf im Browser läuft.
        Membership(
            isPremium: isPremium ?? false,
            premiumEnabled: checkoutAvailable ?? premiumEnabled ?? false,
            limitsActive: limitsActive ?? premiumEnabled ?? false,
            storePurchaseAvailable: storePurchaseAvailable ?? false,
            storeProductID: storeProductId,
            betaActive: betaActive ?? true,
            hasStripeSubscription: hasStripeSubscription ?? isSubscribed ?? false,
            priceCents: priceCents ?? 1000,
            currency: currency ?? "EUR",
            freeDailyLikes: freeDailyLikes ?? 20,
            freeOpenChats: freeOpenChats ?? 3,
            freeMaxRadiusKm: freeMaxRadiusKm ?? 50,
            maxRadiusKm: maxRadiusKm ?? 250,
            likesRemaining: likesRemaining,
            openChatsRemaining: openChatsRemaining,
            nextLikeAt: nextLikeAt.flatMap(ServerTime.parse)
        )
    }
}

extension MessageDTO {
    func toDomain() -> Message {
        Message(
            id: id,
            matchID: matchId,
            senderID: senderId,
            content: content,
            createdAt: ServerTime.parse(createdAt) ?? Date(),
            readAt: ServerTime.parse(readAt),
            wasCensored: wasCensored ?? false
        )
    }
}

extension MatchDTO {
    func toDomain() -> MatchSummary {
        MatchSummary(
            matchID: matchId,
            profile: profile.toDomain(),
            lastMessage: lastMessage?.toDomain(),
            unreadCount: unreadCount ?? 0,
            isOnline: isOnline ?? false,
            inChats: inChats ?? (lastMessage != nil)
        )
    }
}

extension GymDTO {
    func toDomain() -> Gym {
        Gym(
            id: id,
            name: name,
            street: street,
            houseNumber: houseNumber,
            plz: plz,
            city: city,
            label: label
        )
    }
}

extension VerificationStatusDTO {
    func toDomain() -> VerificationState {
        VerificationState(
            status: VerificationStatus(raw: status),
            prompts: prompts ?? [],
            nextStep: VerificationNextStep(raw: nextStep),
            reason: reason,
            verificationRequired: verificationRequired ?? false,
            accountActivated: accountActivated ?? true,
            emailVerified: emailVerified ?? true,
            documentTypes: (documentTypes ?? []).map { $0.toDomain() }
        )
    }
}

extension EmailResendResponseDTO {
    func toDomain() -> EmailResendInfo {
        EmailResendInfo(email: email, validHours: validHours ?? 24)
    }
}

extension EmailConfirmResponseDTO {
    func toDomain() -> EmailConfirmation {
        EmailConfirmation(email: email, name: name, confirmed: confirmed ?? true)
    }
}

extension VerificationDocumentTypeDTO {
    func toDomain() -> VerificationDocumentType {
        VerificationDocumentType(
            value: value,
            label: label,
            needsBack: needsBack ?? false
        )
    }
}

extension PushNotificationDTO {
    func toDomain() -> PushNotification {
        PushNotification(id: id, topic: topic, title: title, body: body, target: target)
    }
}
